src/lib/runner.js

```diff
@@ -1,127 +1,206 @@
-import fs from 'fs';
-import os from 'os';
-import path from 'path';
-import { spawn, spawnSync } from 'child_process';
+import fs from 'node:fs/promises';
+import os from 'node:os';
+import path from 'node:path';
+import { spawn } from 'node:child_process';
 
 class Runner {
-  constructor(enableStdOut = false, forwardParentArgs = false) {
-    this.enableStdOut = enableStdOut; // debugging tests
-    this.forwardParentArgs = forwardParentArgs;
-    this.setupFiles = [];
-    this.childProcess = null;
+  #rootDir;
+  #setupFiles = [];
+  #childProcess = null;
+
+  #enableDebug = false;
+  #forwardParentArgs = false;
+
+  // JS strings can be up to 2^53 - 1 in length,
+  // so no need to worry about overflow
+  // https://262.ecma-international.org/16.0/index.html#sec-ecmascript-language-types-string-type
+  #stdOutBuffer = '';
+  #stdErrBuffer = '';
+
+  #stdOutSubscribers = [];
+  #stdErrSubscribers = [];
+
+  constructor(enableDebug = false, forwardParentArgs = false) {
+    this.#enableDebug = enableDebug; // debugging tests
+    this.#forwardParentArgs = forwardParentArgs;
   }
 
-  setup(rootDir, setupFiles = [], options = { create: true }) {
-    this.setupFiles = setupFiles;
+  async setup(rootDir, setupFiles = null, options = { create: true }) {
+    setupFiles ??= [];
 
-    return new Promise((resolve, reject) => {
-      if (path.isAbsolute(rootDir)) {
-        this.rootDir = rootDir;
+    this.#setRootDir(rootDir);
+    this.#setSetupFiles(setupFiles);
 
-        if (options.create) {
-          fs.mkdirSync(this.rootDir, { recursive: true });
+    if (options.create) {
+      try {
+        await fs.mkdir(this.#rootDir, { recursive: true });
+      } catch (err) {
+        if (err.message?.startsWith('EEXIST: file already exists')) {
```

Comment:
Added this check because apparently `recursive:true` only ignores errors when passed a directory. During testing I still got this error so I'm silencing it because we don't actually care about it.